Feature/dumplibs

ahmed-musallam opened the pull request:
# dumplibs

If you've ever used [http://localhost:4502/libs/granite/ui/content/dumplibs.html](http://localhost:4502/libs/granite/ui/content/dumplibs.html) you know that it works, but the UI is horrible and sometimes you cant find what you're looking for easily. The tool in this PR fixes all that. Most of the functionality in the original `dumplibs.html` but with a much nicer UI!

> Unapologetically Built with [VueJS](https://vuejs.org/), because I don't like angularJs :)

The tool is accessible  by going to [http://localhost:4502/etc/acs-tools/dumplibs.html](http://localhost:4502/etc/acs-tools/dumplibs.html).

### Testing:
I have only tested this on AEM 6.3 due to time constraint, maybe I'll test it on other versions of AEM in the future.

### HTTP API
  - an HTTP API is exposed at `/bin/acs-tools/dumplibs.json`
  - API params:
     - `path`: the clientlib JCR path (takes precedence over `categories` if both are present). returns ONE clientlib
     - `categories` a comma-separated string of the categories to search for. returns ALL clientlibs with categories. 
     - `type`: either `JS` or `CSS`
     - `trans`: (used only when `categories` is specified)[transitive](https://helpx.adobe.com/experience-manager/6-3/sites/developing/using/reference-materials/javadoc/com/adobe/granite/ui/clientlibs/HtmlLibraryManager.html#getLibraries(java.lang.String[],%20com.adobe.granite.ui.clientlibs.LibraryType,%20boolean,%20boolean))
     - `themed`: (used only when `categories` is specified) [themed](https://helpx.adobe.com/experience-manager/6-3/sites/developing/using/reference-materials/javadoc/com/adobe/granite/ui/clientlibs/HtmlLibraryManager.html#getLibraries(java.lang.String[],%20com.adobe.granite.ui.clientlibs.LibraryType,%20boolean,%20boolean)) clientlib
  - If no params specified, all clientlibs are returned

### Features:
  - All clientlibs are loaded on first page load, and then filtered on the front-end
  - You can filter the list of clientlibs via `path`, `categories`, `channels` and `type`. Filters are evaluated with an AND operation.
  - Filters are applied as-you-type :)
  - Clicking on the `path` field will open a dialog showing more details about the clienlib for both `JS` and `CSS` types.
  - Clicking on one of the  `categories` will open a dialog showing ALL clientlibs that use that category for both `JS` and `CSS` types.
  - links with the `pencil` icon will open that link in CRX/DE for easy access. Clicking the link itself will take you to that link directly.

### Screenshots:
<img width="1440" alt="screen shot 2018-06-24 at 9 36 30 pm" src="https://user-images.githubusercontent.com/13037215/41827759-e465a92c-77f6-11e8-8147-a852936c257e.png">
<img width="1418" alt="screen shot 2018-07-20 at 12 02 31 am" src="https://user-images.githubusercontent.com/13037215/42984086-61a0a528-8bb0-11e8-8329-67141a69813a.png">
<img width="1418" alt="screen shot 2018-07-20 at 12 03 04 am" src="https://user-images.githubusercontent.com/13037215/42984088-61b37a40-8bb0-11e8-8e4e-95ec69b1b6ae.png">


### Future enhancements 
  - Add output testing, libraries validation and libraries rebuild akin the original dumplibs.

> PS: I know I should have done a better a job with breaking my commits into smaller chunks... but I got too excited and forgot to commit as often as I usually do :) this was a one-weekend project anyway :)

muhammad-naveed commented on bundle/src/main/java/com/adobe/acs/tools/dumplibs/DumplibsParams.java:
```diff
@@ -0,0 +1,82 @@
+        type = getLibraryType(request.getParameter(PARAM_TYPE));
+    }
+
+    private  String[] getCatigoriesArray(String catigs){
```
There is a typo in the method name. It should be getCategoriesArray() instead of getCatigoriesArray()

ahmed-musallam replied:
@muhammad-naveed good catch! I keep typing 1categories` as `catigories` for some reason, and I keep going back to fix it.. smh

Anyway, fixed it!

davidjgonzalez commented:
Thanks! Should have time to check it out this week!!

davidjgonzalez commented on bundle/src/main/java/com/adobe/acs/tools/dumplibs/DumplibsServlet.java:
```diff
@@ -0,0 +1,217 @@
+        label = "ACS AEM Tools - Dumplibs",
+        description = "ACS AEM Tools End-point for Dumplibs",
+        methods = {"GET"},
+        paths = {"/bin/acs-tools/dumplibs"},
```
This should be bound to the resource type of `/etc/acs-tools/dumplibs` and use that resource as the entry point. Generally speaking, path-based servlets are an anti-pattern.

If needed you can have this response to a selector to isolate this servlets behavior.

ahmed-musallam replied:
@davidjgonzalez sure, I can make that change

ahmed-musallam replied:
@davidjgonzalez actually, I don't see the issue with using path based servlet for this purpose. This servlet has nothing to do with the resource itself. The servlet calls the `HtmlLibraryManager` service exclusively and uses nothing of the resource, or any other resources for that matter. As such I think the use of path here is warranted and appropriate.

If your `anti-pattern` argument was referencing this: [https://sling.apache.org/documentation/the-sling-engine/servlets.html#caveats-when-binding-servlets-by-path](https://sling.apache.org/documentation/the-sling-engine/servlets.html#caveats-when-binding-servlets-by-path); none of the caveats apply in this case.

Let me know your thoughts, 
thanks!

davidjgonzalez replied:
using the path-based approach introduces a new (virtual) unneeded virtual path that is also (more) difficult to permission.

Basically, there's no good reason TO introduce this path. You already have a resource (/etc/acs-tools/dumplibs) that represents your tool, which is what the HTTP requests (that are the actions for this tool) should interact with... 

The servlet has everything to do w/ the resource since the resource represents your tool (and to be clear your tool != HtmlLibraryManager). Most servlet end-points just call other Services to do the work (or atleast they probably should), but they still are bound to resources.

Hope that helps clarify?

ahmed-musallam replied:
@davidjgonzalez, while I disagree on the point regarding this being a "unneeded virtual path" I do agree on the permission point, even thought not extremely relevant in this specific case.

I *dont mind*  registering this servlet with a `resouceType`, and to avoid further discussion and delays I've implemented it, please see my latest commits :)

> I still prefer the path, in this very specific case, as it is more explicit for an external tool and other reasons I mentioned above, but that's just me :)  `resouceType` based servlets are usually the way to go, when it makes sense, which is most cases, but this isn't most cases :)

davidjgonzalez commented on bundle/src/main/java/com/adobe/acs/tools/dumplibs/DumplibsServlet.java:
```diff
@@ -0,0 +1,217 @@
+        HtmlLibrary lib = libraryManager.getLibrary(p.getType(), p.getPath());
+        try {
+
+            JSONObject libJSON = htmlLibraryToJSON(lib); // should never be null
```
Sling JSON is deprecated :( .. were moving over to GSON (which is provided by AEM) or Sling Model Exporter + Jackson .. i assume this will be easier to switch to GSON than SME.

ahmed-musallam replied:
@davidjgonzalez I dont know that GSON is available in older AEM versions and I do know that Sling JSON is deprecated, I chose it for backwards compatibility and because other code in this repo uses it. 

So how do you think we should handle including GSON dependency? Should this project include GSON? should it be provided? if so which version of GSON is included in which versions of AEM?

davidjgonzalez replied:
ACS Tools now only supported on AEM 6.2+ -- which i'm fairly certain includes the GSON dependency (since that's what we've moved to for ACS Commons which also supports 6.2+)

You can use the same as what's used in ACS Commons and we should be good (and future proof).
https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/bundle/pom.xml#L247

ahmed-musallam replied:
@davidjgonzalez moved to GSON as per your recommendation, see my latest commits :)

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-tools/components/dumplibs/clientlibs/js/components/clientlib-modal.js:
```diff
@@ -0,0 +1,115 @@
+        },
+        methods: {
+            getClientlibs: function (params) {
+                return axios.get('/bin/acs-tools/dumplibs.json', {
```
This should change to `/etc/acs-tools/dumplibs.xxx.json` ... ideally, it would support custom servlet contexts, but not sure if there's a simple/clean way to expose that to the Vue app. If not, we should just make sure to document it doesn't support custom contexts..

ahmed-musallam replied:
I dont know what you mean by `custom servlet contexts` ?

davidjgonzalez replied:
If you run aem on http://host:port/custom-context/etc/acs-tools/dumplibs.html

ahmed-musallam replied:
@davidjgonzalez still not getting what you mean? is this a sling/java servlet concept that I've never heard of? or is it some kind of pattern? Also, are there an examples I can reference?

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-tools/components/dumplibs/clientlibs/js/components/crx-link.js:
```diff
@@ -0,0 +1,40 @@
+                return Vue.type.isString(l) && l.startsWith('/');
+            },
+            crxLink: function (link) {
+                return '/crx/de/index.jsp#' + link;
```
Same as above wrt to servlet context path ... idk if this is a total PITA or easy.

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-tools/components/dumplibs/content.jsp:
```diff
@@ -0,0 +1,25 @@
+  --%>
+
+<%@include file="/libs/foundation/global.jsp" %>
+<div id="acs-tools-dumplibs">
```
could the servlet context be provided via a data-* here? Or is that an anti-pattern for vue?

davidjgonzalez commented on pom.xml:
```diff
@@ -409,6 +409,15 @@
         Repository Manager. Once you do that, remove these sections.
      -->
     <repositories>
+        <repository>
```
This should be removed.

ahmed-musallam replied:
ahh yes, good catch! will remove that

davidjgonzalez commented on pom.xml:
```diff
@@ -417,6 +426,18 @@
         </repository>
     </repositories>
     <pluginRepositories>
+        <pluginRepository>
```
This should be removed.

ahmed-musallam replied:
ahh yes, good catch! will remove that

ahmed-musallam commented:
@davidjgonzalez I am not sure what you mean by servlet context? could you expand more?
I do get the points about GSON and path servlet and I had questions, so I replied to your code review comments.

ahmed-musallam commented:
@davidjgonzalez I made some of the changes you requested:
1. moved HTTP requests to a central Vue plugin
2. removed the `repositories` that were added by mistake

There are some I still have questions about, replied to them indivisually
Additional changes:
1. helpful messages in dialog when clientlib cannot be found
2. documentation for my JS code

ahmed-musallam commented:
Hey @davidjgonzalez any chance you looked at my comments/changes?

ahmed-musallam commented:
@davidjgonzalez reminder: I've made the requested changes, please review and let me know.

davidjgonzalez commented:
@ahmed-musallam looks great! I PR'd a small change (to the end-point) to your branch - if you can merge the PR to your branch, then i can merge through to master!

Really nice code btw! 👍 🥇

ahmed-musallam commented:
@davidjgonzalez sure, merged

kwin commented:
@ahmed-musallam Really nice tool. Would it be possible to also expose the dependencies and embedded clientlibs per row?

ahmed-musallam commented:
@kwin I add this a while ago. I remember there was an API reason I could not expose `dependencies` and `embed` categories.. but I don't remember the exact reason.. 😅. 

If I have some time, I'll look into it and document it.
